marimbabot_audio/scripts/onset_detection.py

```diff
@@ -0,0 +1,547 @@
+            output:
+                -   winner_freq: the freq of the winner signal
+                -   max(buckets[winner]): the confidence
+                -   winner_idx_in_spec: the idx in spec along y-axis
+                -   winner_pm_idx: the note id in the pretty_midi
+        """
+        # using 0.1 sec windows data for classification
+        prediction_averaging_window = (
+            self.windows_for_classification * self.sr
+        )
+        # extract the data from onset time until 0.1 sec later
+        excerpt = self.buffer[int(onset * self.sr):int(onset * self.sr + prediction_averaging_window)]
+        # neuron nets for onset classification
+        time, freq, confidence, _ = crepe.predict(
+            excerpt,
+            self.sr,
+            viterbi=True,
+            model_capacity=self.crepe_model,
+            verbose=0
+        )
+
+        # filter out the onset, which confidence lower that threshold and note beyond range(C4-C7)
+        confidence_mask = confidence > self.confidence_threshold
+        freq_mask = (freq >= self.fmin) & (freq <= self.fmax)
+        mask = confidence_mask & freq_mask
+        filtered_freq = freq[mask]
+        filtered_confidence = confidence[mask]
+
+        if len(filtered_freq) > 0:
+            buckets = {}
+            for f, c in zip(filtered_freq, filtered_confidence):
+                note = hz_to_note(f)
+                buckets[note] = buckets.get(note, []) + [c]
+
+            def add_confidence(note):
+                return reduce(lambda x, y: x + y, buckets.get(note))
+            winner = max(buckets, key=lambda a: add_confidence(a))
+            winner_freq = note_to_hz(winner)
+            winner_pm_idx = note_to_pm_id(winner)
+            winner_idx_in_spec = note_to_pm_id(winner)-self.fmin_note_id
+
+            return winner_freq, max(buckets[winner]), winner_idx_in_spec, winner_pm_idx
+        else:
+            return 0.0, 0.0, None, None
+
+    def update_spectrogram(self, spec, ys, xs, durations_len):
+        if self.pub_spectrogram.get_num_connections() == 0:
+            self.spectrogram = None
+            return
+
+        # throw away overlap
+        spec = spec[:, self.overlap_hops:-self.overlap_hops]
+
+        log_spec = np.log(spec+1e-8)
+
+        if self.spectrogram is None:
+            self.spectrogram = log_spec
+            return
+        elif self.spectrogram.shape[1] > spec.shape[1]:
+            self.spectrogram = self.spectrogram[:, -spec.shape[1]:]
+        self.spectrogram = np.concatenate([self.spectrogram, log_spec], 1)
+
+        # normalizes per compute
+        spectrogram = np.array(
+            self.spectrogram / np.max(self.spectrogram) * 255, dtype=np.uint8
+        )
+
+        heatmap = cv2.applyColorMap(spectrogram, cv2.COLORMAP_JET)
+
+        assert len(durations_len) == len(xs)
+
+        LINECOLOR = [255, 0, 255]
+        for idx, t in enumerate(self.previous_onsets_times):
```

Comment:
You should be able to vectorize this without a for loop using only numpy.

Reply:
later on when speed not satisfied.